File: app-ios/tutanota/Sources/Utils/AsyncBlockOperation.m
//
//  AsyncBlockOperation.m
//  tutanota
//
//  Created by Tutao GmbH on 18.02.20.
//  Copyright © 2020 Tutao GmbH. All rights reserved.
//

#import "AsyncBlockOperation.h"

@interface AsyncBlockOperation () {
    BOOL _finished;
    BOOL _executing;
}

@property (nonatomic, copy) AsyncBlock block;

@end


@implementation AsyncBlockOperation

+ (instancetype)asyncBlockOperationWithBlock:(AsyncBlock)block {
    return [[AsyncBlockOperation alloc] initWithAsyncBlock:block];
}

- (instancetype)initWithAsyncBlock:(AsyncBlock)block {
    if (self = [super init]) {
        self.block = block;
    }
    return self;
}

- (void)start {
    [self willChangeValueForKey:@"isExecuting"];
    _executing = YES;
    [self didChangeValueForKey:@"isExecuting"];
    
    self.block(^{
        [self willChangeValueForKey:@"isExecuting"];
        self->_executing = NO;
        [self didChangeValueForKey:@"isExecuting"];
        [self willChangeValueForKey:@"isFinished"];
        self->_finished = YES;
        [self didChangeValueForKey:@"isFinished"];
    });
}

- (BOOL)isFinished {
    return _finished;
}

- (BOOL)isExecuting {
    return _executing;
}

- (BOOL)isAsynchronous {
    return YES;
}

@end

@implementation NSOperationQueue (AsyncBlockOperation)

- (void)addAsyncOperationWithBlock:(AsyncBlock)block {
    [self addOperation:[AsyncBlockOperation asyncBlockOperationWithBlock:block]];
}

@end

File: app-ios/tutanota/Sources/Utils/AsyncBlockOperation.swift
import Foundation

typealias AsyncBlock = (@escaping () -> Void) -> Void

class AsyncBlockOperation : Foundation.Operation {
  // We need to keep KVO and it's hard to do with enum
  private var _isExecuting: Bool {
    willSet {
      self.willChangeValue(forKey: "isExecuting")
    }
    didSet {
      self.didChangeValue(forKey: "isExecuting")
    }
  }
  
  private var _isFinished: Bool {
    willSet {
      self.willChangeValue(forKey: "isFinished")
    }
    didSet {
      self.didChangeValue(forKey: "isFinished")
    }
  }
  
  private let block: AsyncBlock
  
  init(block: @escaping AsyncBlock) {
    self.block = block
    self._isExecuting = false
    self._isFinished = false
    super.init()
  }
  
  @objc
  override func start() {
    self._isExecuting = true
    self.block {
      self._isExecuting = false
      self._isFinished = true
    }
  }
  
  override var isExecuting: Bool {
    return self._isExecuting
  }
  
  override var isFinished: Bool {
    return self._isFinished
  }
  
  @objc
  override var isAsynchronous: Bool {
    get {
      return true
    }
  }
}

extension OperationQueue {
  func addAsyncOperation(block: @escaping AsyncBlock) {
    self.addOperation(AsyncBlockOperation(block: block))
  }
}
